src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+        let id = self.generate_id();
+        instruction.set_result(id);
+
+        // TODO Support other imports
+        instruction.add_operands(self.string_to_words("GLSL.std.450"));
+        self.lookup_import.insert(id, String::from("GLSL.std.450"));
+
+        instruction
+    }
+
+    fn instruction_memory_model(&mut self) -> Instruction {
+        let mut instruction = Instruction::new(Op::MemoryModel);
+        let addressing_model = AddressingModel::Logical;
+        let memory_model = MemoryModel::GLSL450;
+        self.try_add_capabilities(addressing_model.required_capabilities());
+        self.try_add_capabilities(memory_model.required_capabilities());
+
+        instruction.add_operand(addressing_model as u32);
+        instruction.add_operand(memory_model as u32);
+        instruction
+    }
+
+    fn instruction_entry_point(
+        &mut self,
+        entry_point: &crate::EntryPoint,
+        ir_module: &crate::Module,
+    ) -> Instruction {
+        let mut instruction = Instruction::new(Op::EntryPoint);
+        let function_id = self
+            .lookup_function
+            .lookup_id(&entry_point.function)
+            .unwrap();
+
+        instruction.add_operand(entry_point.exec_model as u32);
+        instruction.add_operand(function_id);
+
+        if self.debug_enabled {
+            let mut debug_instruction = Instruction::new(Op::Name);
+            debug_instruction.set_result(function_id);
+            debug_instruction.add_operands(self.string_to_words(entry_point.name.as_str()));
+            self.debugs.push(debug_instruction);
+        }
+
+        instruction.add_operands(self.string_to_words(entry_point.name.as_str()));
+
+        let function = &ir_module.functions[entry_point.function];
+        for ((handle, _), &usage) in ir_module
+            .global_variables
+            .iter()
+            .zip(&function.global_usage)
+        {
+            if usage.contains(crate::GlobalUse::STORE) || usage.contains(crate::GlobalUse::LOAD) {
+                let id = self.get_global_variable_id(
+                    &ir_module.types,
+                    &ir_module.global_variables,
+                    &handle,
+                );
+                instruction.add_operand(id);
+            }
+        }
+
+        self.try_add_capabilities(entry_point.exec_model.required_capabilities());
+        match entry_point.exec_model {
+            ExecutionModel::Vertex | ExecutionModel::GLCompute => {}
+            ExecutionModel::Fragment => {
+                let execution_mode = ExecutionMode::OriginUpperLeft;
+                self.try_add_capabilities(execution_mode.required_capabilities());
+                let mut execution_mode_instruction = Instruction::new(Op::ExecutionMode);
+                execution_mode_instruction.add_operand(function_id);
+                execution_mode_instruction.add_operand(execution_mode as u32);
+                for word in execution_mode_instruction.to_words() {
```

Comment:
In most, if not all, of these places, the heap allocation in `to_words` isn't needed. One way to get around is by using the `Extend<Word>` generic